clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java

```diff
@@ -143,6 +146,39 @@ private void setupCoordinator(int retryBackoffMs, int rebalanceTimeoutMs, Option
                                                 mockTime);
     }
 
+    @Test
+    public void testMetrics() {
```

Comment:
In this test, you verify if you register the metrics with the correct name, but you do not verify if the `MeasureableStat`s are correct. Similar is true for the `ConsumerCoordinatorTest`.  You can find an example of how to verify that in `StreamsMetricsImplTest#shouldAddAmountRateAndSum()`.

Reply:
I think I agree with you. But looking at the other unit tests I think instead of adding the logic of test recording behavior in a single `testMetrics` we'd better adding the logic in each individual test case, e.g. during different rebalance scenarios test failed-rebalance and callback trigger rates, etc. And it should be a rather big effort adding this coverage in different modules.

How about I create a JIRA ticket for doing this as a separate task?

Reply:
I am not sure if I completely follow what your are saying. Looking again at the code, I think, it would suffice to pass a mock `Metrics` object to `AbstractCoordinator`. The mock would return a `Sensor` mock for a call to `sensor()`, with which you can verify if the parameter passed to `add()` is an object of class `Avg` (or whatever `MeasurableStat` is used). This should be possible with `PowerMock`. We have to use `PowerMock` since `Sensor` is `final`.

If you think, it is too much work given the upcoming release, I am fine with creating a Jira and postpone.

Reply:
Yes, I'm thinking about using `MockMetricsReporter`, e.g. the pattern in `KafkaStreamsTest#testInitializesAndDestroysMetricsReporters`. What I was saying is that I'd like to just add this checking logic in each of the existing test function than doing that in a single `testMetrics` function for all scenarios, e.g. different patterns of rebalance failures where `failed-rebalance` should be recorded -- basically each test function within `ConsumerCoordinatorTest` that described a normal or abnormal path. And looking at other modules that have an embedded metrics model, we actually do not test whether the recording behavior is expected as-is or not in most of them, and I think having a single task of adding such test coverage on all such modules by:

1) using a mock metrics reporter,
2) check the corresponding metrics are correctly recorded under different scenarios.

And it would touch on Selector, Sender, ConsumerNetworkClient, Fetcher, Abstract/ConsumerCoordinator, NetworkClient which all have a corresponding `XXXTest` but they do not have metrics recording in place.